File: apps/native/macos/Sources/Icons.swift
import AppKit

/// Stroke icons on a 16 pt grid, matching `Icon.tsx` and the egui host.
enum Icon {
    case plus, more, database, refresh, table, key, view, folder
    case chevronRight, chevronDown, chevronUp, chevronLeft
    case play, close, copy, download, inspector, check, alert, code, sortUp, sortDown, search, trash, undo, sidebar
    case pencil, expand, collapse, filter

    /// Draws into the current context, in a flipped 16×16 space scaled to `rect`.
    func draw(in rect: NSRect, color: NSColor) {
        let scale = min(rect.width, rect.height) / 16
        let origin = NSPoint(x: rect.midX - 8 * scale, y: rect.midY - 8 * scale)
        let flipped = NSGraphicsContext.current?.isFlipped ?? true
        func p(_ x: CGFloat, _ y: CGFloat) -> NSPoint {
            let py = flipped ? origin.y + y * scale : origin.y + (16 - y) * scale
            return NSPoint(x: origin.x + x * scale, y: py)
        }
        func r(_ x: CGFloat, _ y: CGFloat, _ w: CGFloat, _ h: CGFloat) -> NSRect {
            let a = p(x, y), b = p(x + w, y + h)
            return NSRect(x: min(a.x, b.x), y: min(a.y, b.y), width: abs(b.x - a.x), height: abs(b.y - a.y))
        }
        color.setStroke()
        color.setFill()
        let path = NSBezierPath()
        path.lineWidth = 1.5 * max(scale, 0.75)
        path.lineCapStyle = .round
        path.lineJoinStyle = .round
        func line(_ points: NSPoint...) {
            path.move(to: points[0])
            points.dropFirst().forEach { path.line(to: $0) }
        }
        func dot(_ x: CGFloat, _ y: CGFloat, _ radius: CGFloat) {
            let center = p(x, y)
            NSBezierPath(ovalIn: NSRect(x: center.x - radius * scale, y: center.y - radius * scale,
                                        width: 2 * radius * scale, height: 2 * radius * scale)).fill()
        }
        func fillPolygon(_ points: NSPoint...) {
            let shape = NSBezierPath()
            shape.move(to: points[0])
            points.dropFirst().forEach { shape.line(to: $0) }
            shape.close()
            shape.fill()
        }
        switch self {
        case .plus: line(p(8, 3), p(8, 13)); line(p(3, 8), p(13, 8))
        case .more: [3.5, 8, 12.5].forEach { dot($0, 8, 1.2) }
        case .database:
            path.appendOval(in: r(3, 2.2, 10, 3.6))
            line(p(3, 4), p(3, 12)); line(p(13, 4), p(13, 12))
            for y: CGFloat in [8, 12] {
                path.move(to: p(3, y))
                path.curve(to: p(13, y), controlPoint1: p(3, y + 2.4), controlPoint2: p(13, y + 2.4))
            }
        case .refresh:
            // Grid coordinates grow downward, as in the egui host.
            let points = (0...20).map { i -> (CGFloat, CGFloat) in
                let angle = (-40 + CGFloat(i) * 14) * .pi / 180
                return (8 + 4.8 * cos(angle), 8 + 4.8 * sin(angle))
            }
            path.move(to: p(points[0].0, points[0].1))
            points.dropFirst().forEach { path.line(to: p($0.0, $0.1)) }
            let end = points[points.count - 1]
            line(p(end.0, end.1), p(end.0 + 0.4, end.1 - 3))
            line(p(end.0, end.1), p(end.0 + 3, end.1 - 0.4))
        case .table:
            path.appendRoundedRect(r(2.5, 3, 11, 10), xRadius: 1.5 * scale, yRadius: 1.5 * scale)
            line(p(2.5, 6.5), p(13.5, 6.5)); line(p(6.5, 6.5), p(6.5, 13))
        case .key:
            path.appendOval(in: r(2.9, 5.4, 5.2, 5.2))
            line(p(8.1, 8), p(13.5, 8)); line(p(11.5, 8), p(11.5, 10.5)); line(p(13.5, 8), p(13.5, 10))
        case .view:
            path.appendRoundedRect(r(2.5, 3, 11, 10), xRadius: 1.5 * scale, yRadius: 1.5 * scale)
            line(p(2.5, 6.5), p(13.5, 6.5))
            path.appendOval(in: r(6.4, 8.2, 3.2, 3.2))
        case .folder:
            line(p(2.5, 4.5), p(6.5, 4.5), p(7.8, 6), p(13.5, 6), p(13.5, 12.5), p(2.5, 12.5), p(2.5, 4.5))
        case .chevronRight: line(p(6, 4), p(10, 8), p(6, 12))
        case .chevronLeft: line(p(10, 4), p(6, 8), p(10, 12))
        case .chevronDown: line(p(4, 6), p(8, 10), p(12, 6))
        case .chevronUp: line(p(4, 10), p(8, 6), p(12, 10))
        case .play: fillPolygon(p(5, 3.5), p(12.5, 8), p(5, 12.5))
        case .close: line(p(4.5, 4.5), p(11.5, 11.5)); line(p(11.5, 4.5), p(4.5, 11.5))
        case .copy:
            path.appendRoundedRect(r(5.5, 5.5, 7.5, 7.5), xRadius: 1.5 * scale, yRadius: 1.5 * scale)
            line(p(3, 10.5), p(3, 3), p(10.5, 3))
        case .download: line(p(8, 2.5), p(8, 10)); line(p(4.5, 7), p(8, 10.5), p(11.5, 7)); line(p(3, 13), p(13, 13))
        case .inspector:
            path.appendRoundedRect(r(2.5, 3, 11, 10), xRadius: 1.5 * scale, yRadius: 1.5 * scale)
            line(p(9.5, 3), p(9.5, 13))
        case .sidebar:
            path.appendRoundedRect(r(2.5, 3, 11, 10), xRadius: 1.5 * scale, yRadius: 1.5 * scale)
            line(p(6.5, 3), p(6.5, 13))
        case .check: line(p(3.5, 8.5), p(6.5, 11.5), p(12.5, 4.5))
        case .alert:
            line(p(8, 2.5), p(14, 13), p(2, 13), p(8, 2.5))
            line(p(8, 6.5), p(8, 9.2)); dot(8, 11.1, 0.8)
        case .code: line(p(6, 4.5), p(2.5, 8), p(6, 11.5)); line(p(10, 4.5), p(13.5, 8), p(10, 11.5))
        case .sortUp: fillPolygon(p(8, 5), p(11.5, 10.5), p(4.5, 10.5))
        case .sortDown: fillPolygon(p(4.5, 5.5), p(11.5, 5.5), p(8, 11))
        case .search: path.appendOval(in: r(3, 3, 7.5, 7.5)); line(p(9.3, 9.3), p(13, 13))
        case .filter: line(p(2, 3.3), p(14, 3.3), p(9.3, 8.7), p(9.3, 12.7), p(6.7, 11.3), p(6.7, 8.7), p(2, 3.3))
        case .trash:
            line(p(3, 4.5), p(13, 4.5)); line(p(6.5, 4.5), p(6.5, 3), p(9.5, 3), p(9.5, 4.5))
            line(p(4.5, 4.5), p(5.2, 13), p(10.8, 13), p(11.5, 4.5))
        case .pencil:
            line(p(3, 13), p(3.6, 10.2), p(10.8, 3), p(13, 5.2), p(5.8, 12.4), p(3, 13))
            line(p(9.3, 4.5), p(11.5, 6.7))
        case .expand:
            line(p(3, 8), p(13, 8)); line(p(10, 5), p(13, 8), p(10, 11)); line(p(3, 4), p(3, 12))
        case .collapse:
            line(p(3, 8), p(13, 8)); line(p(6, 5), p(3, 8), p(6, 11)); line(p(13, 4), p(13, 12))
        case .undo:
            line(p(5.5, 3.5), p(3, 6), p(5.5, 8.5))
            path.move(to: p(3, 6))
            path.curve(to: p(8, 13), controlPoint1: p(12, 6), controlPoint2: p(14, 13))
        }
        path.stroke()
    }

    func image(size: CGFloat = 14, color: NSColor) -> NSImage {
        let image = NSImage(size: NSSize(width: size, height: size), flipped: true) { rect in
            self.draw(in: rect, color: color)
            return true
        }
        image.isTemplate = false
        return image
    }
}

/// A non-interactive icon.
final class IconView: NSView {
    var icon: Icon { didSet { needsDisplay = true } }
    var color: NSColor { didSet { needsDisplay = true } }

    init(_ icon: Icon, size: CGFloat = 14, color: NSColor = Graphite.muted) {
        self.icon = icon
        self.color = color
        super.init(frame: NSRect(x: 0, y: 0, width: size, height: size))
        translatesAutoresizingMaskIntoConstraints = false
        widthAnchor.constraint(equalToConstant: size).isActive = true
        heightAnchor.constraint(equalToConstant: size).isActive = true
    }

    required init?(coder: NSCoder) { fatalError("init(coder:) is not supported") }

    override var isFlipped: Bool { true }

    override func draw(_ dirtyRect: NSRect) {
        icon.draw(in: bounds, color: color)
    }
}

File: apps/native/workbench/src/icons.rs
//! Stroke icons painted on a 16 px grid, matching the 1.6 px glyphs in the
//! desktop app's `Icon.tsx` without depending on font coverage.

use eframe::egui::{self, Color32, Pos2, Rect, Response, Sense, Shape, Stroke, Vec2};

use crate::theme;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Icon {
    Plus,
    More,
    Database,
    Refresh,
    Table,
    Key,
    View,
    ChevronRight,
    ChevronDown,
    ChevronUp,
    ChevronLeft,
    Play,
    Close,
    Copy,
    Download,
    Inspector,
    Sidebar,
    Pencil,
    Expand,
    Collapse,
    Check,
    Alert,
    Code,
    SortUp,
    SortDown,
    Folder,
    Trash,
    Undo,
    Filter,
}

/// Paints `icon` centered in `rect` using a 16 px design grid.
pub fn paint(painter: &egui::Painter, rect: Rect, icon: Icon, color: Color32) {
    let scale = rect.width().min(rect.height()) / 16.0;
    let origin = rect.center() - Vec2::splat(8.0 * scale);
    let p = |x: f32, y: f32| Pos2::new(origin.x + x * scale, origin.y + y * scale);
    let stroke = Stroke::new(1.5 * scale.max(0.75), color);
    let line = |a: Pos2, b: Pos2| {
        painter.line_segment([a, b], stroke);
    };
    let path = |points: Vec<Pos2>| {
        painter.add(Shape::line(points, stroke));
    };
    match icon {
        Icon::Plus => {
            line(p(8.0, 3.0), p(8.0, 13.0));
            line(p(3.0, 8.0), p(13.0, 8.0));
        }
        Icon::More => {
            for x in [3.5, 8.0, 12.5] {
                painter.circle_filled(p(x, 8.0), 1.2 * scale, color);
            }
        }
        Icon::Database => {
            painter.add(egui::epaint::EllipseShape::stroke(
                p(8.0, 4.0),
                Vec2::new(5.0, 1.8) * scale,
                stroke,
            ));
            line(p(3.0, 4.0), p(3.0, 12.0));
            line(p(13.0, 4.0), p(13.0, 12.0));
            path(arc(p(8.0, 8.0), 5.0 * scale, 1.8 * scale));
            path(arc(p(8.0, 12.0), 5.0 * scale, 1.8 * scale));
        }
        Icon::Refresh => {
            let center = p(8.0, 8.0);
            let radius = 4.8 * scale;
            let points: Vec<Pos2> = (0..=20)
                .map(|i| {
                    let angle = (-40.0 + i as f32 * 14.0_f32).to_radians();
                    center + Vec2::angled(angle) * radius
                })
                .collect();
            let end = *points.last().unwrap_or(&center);
            path(points);
            line(end, end + Vec2::new(0.4, -3.0) * scale);
            line(end, end + Vec2::new(3.0, -0.4) * scale);
        }
        Icon::Table => {
            painter.rect_stroke(
                Rect::from_min_max(p(2.5, 3.0), p(13.5, 13.0)),
                1.5 * scale,
                stroke,
                egui::StrokeKind::Middle,
            );
            line(p(2.5, 6.5), p(13.5, 6.5));
            line(p(6.5, 6.5), p(6.5, 13.0));
        }
        Icon::Key => {
            painter.circle_stroke(p(5.5, 8.0), 2.6 * scale, stroke);
            line(p(8.1, 8.0), p(13.5, 8.0));
            line(p(11.5, 8.0), p(11.5, 10.5));
            line(p(13.5, 8.0), p(13.5, 10.0));
        }
        Icon::Filter => {
            painter.add(Shape::closed_line(
                vec![
                    p(2.0, 3.3),
                    p(14.0, 3.3),
                    p(9.3, 8.7),
                    p(9.3, 12.7),
                    p(6.7, 11.3),
                    p(6.7, 8.7),
                ],
                stroke,
            ));
        }
        Icon::Trash => {
            line(p(3.0, 4.5), p(13.0, 4.5));
            path(vec![p(6.5, 4.5), p(6.5, 3.0), p(9.5, 3.0), p(9.5, 4.5)]);
            path(vec![p(4.5, 4.5), p(5.2, 13.0), p(10.8, 13.0), p(11.5, 4.5)]);
        }
        Icon::Undo => {
            path(vec![p(5.5, 3.5), p(3.0, 6.0), p(5.5, 8.5)]);
            painter.add(egui::epaint::CubicBezierShape::from_points_stroke(
                [p(3.0, 6.0), p(12.0, 6.0), p(14.0, 13.0), p(8.0, 13.0)],
                false,
                Color32::TRANSPARENT,
                stroke,
            ));
        }
        Icon::Folder => {
            painter.add(Shape::closed_line(
                vec![
                    p(2.0, 4.0),
                    p(6.0, 4.0),
                    p(7.5, 5.5),
                    p(14.0, 5.5),
                    p(14.0, 12.5),
                    p(2.0, 12.5),
                ],
                stroke,
            ));
        }
        Icon::View => {
            painter.rect_stroke(
                Rect::from_min_max(p(2.5, 3.0), p(13.5, 13.0)),
                1.5 * scale,
                stroke,
                egui::StrokeKind::Middle,
            );
            line(p(2.5, 6.5), p(13.5, 6.5));
            painter.circle_stroke(p(8.0, 9.8), 1.6 * scale, stroke);
        }
        Icon::ChevronRight => {
            path(vec![p(6.0, 4.0), p(10.0, 8.0), p(6.0, 12.0)]);
        }
        Icon::ChevronLeft => {
            path(vec![p(10.0, 4.0), p(6.0, 8.0), p(10.0, 12.0)]);
        }
        Icon::ChevronUp => {
            path(vec![p(4.0, 10.0), p(8.0, 6.0), p(12.0, 10.0)]);
        }
        Icon::ChevronDown => {
            path(vec![p(4.0, 6.0), p(8.0, 10.0), p(12.0, 6.0)]);
        }
        Icon::Play => {
            painter.add(Shape::convex_polygon(
                vec![p(5.0, 3.5), p(12.5, 8.0), p(5.0, 12.5)],
                color,
                Stroke::NONE,
            ));
        }
        Icon::Close => {
            line(p(4.5, 4.5), p(11.5, 11.5));
            line(p(11.5, 4.5), p(4.5, 11.5));
        }
        Icon::Copy => {
            painter.rect_stroke(
                Rect::from_min_max(p(5.5, 5.5), p(13.0, 13.0)),
                1.5 * scale,
                stroke,
                egui::StrokeKind::Middle,
            );
            path(vec![p(3.0, 10.5), p(3.0, 3.0), p(10.5, 3.0)]);
        }
        Icon::Download => {
            line(p(8.0, 2.5), p(8.0, 10.0));
            path(vec![p(4.5, 7.0), p(8.0, 10.5), p(11.5, 7.0)]);
            line(p(3.0, 13.0), p(13.0, 13.0));
        }
        Icon::Pencil => {
            path(vec![
                p(3.0, 13.0),
                p(3.6, 10.2),
                p(10.8, 3.0),
                p(13.0, 5.2),
                p(5.8, 12.4),
                p(3.0, 13.0),
            ]);
            line(p(9.3, 4.5), p(11.5, 6.7));
        }
        Icon::Expand => {
            line(p(3.0, 8.0), p(13.0, 8.0));
            path(vec![p(10.0, 5.0), p(13.0, 8.0), p(10.0, 11.0)]);
            line(p(3.0, 4.0), p(3.0, 12.0));
        }
        Icon::Collapse => {
            line(p(3.0, 8.0), p(13.0, 8.0));
            path(vec![p(6.0, 5.0), p(3.0, 8.0), p(6.0, 11.0)]);
            line(p(13.0, 4.0), p(13.0, 12.0));
        }
        Icon::Sidebar => {
            painter.rect_stroke(
                Rect::from_min_max(p(2.5, 3.0), p(13.5, 13.0)),
                1.5 * scale,
                stroke,
                egui::StrokeKind::Middle,
            );
            line(p(6.5, 3.0), p(6.5, 13.0));
        }
        Icon::Inspector => {
            painter.rect_stroke(
                Rect::from_min_max(p(2.5, 3.0), p(13.5, 13.0)),
                1.5 * scale,
                stroke,
                egui::StrokeKind::Middle,
            );
            line(p(9.5, 3.0), p(9.5, 13.0));
        }
        Icon::Check => {
            path(vec![p(3.5, 8.5), p(6.5, 11.5), p(12.5, 4.5)]);
        }
        Icon::Alert => {
            painter.circle_stroke(p(8.0, 8.0), 5.5 * scale, stroke);
            line(p(8.0, 5.0), p(8.0, 8.8));
            painter.circle_filled(p(8.0, 11.0), 0.9 * scale, color);
        }
        Icon::Code => {
            path(vec![p(6.0, 4.5), p(2.5, 8.0), p(6.0, 11.5)]);
            path(vec![p(10.0, 4.5), p(13.5, 8.0), p(10.0, 11.5)]);
        }
        Icon::SortUp => {
            painter.add(Shape::convex_polygon(
                vec![p(8.0, 5.0), p(11.5, 10.5), p(4.5, 10.5)],
                color,
                Stroke::NONE,
            ));
        }
        Icon::SortDown => {
            painter.add(Shape::convex_polygon(
                vec![p(4.5, 5.5), p(11.5, 5.5), p(8.0, 11.0)],
                color,
                Stroke::NONE,
            ));
        }
    }
}

/// The lower half of an ellipse, for the database glyph's bands.
fn arc(center: Pos2, rx: f32, ry: f32) -> Vec<Pos2> {
    (0..=16)
        .map(|i| {
            let t = std::f32::consts::PI * i as f32 / 16.0;
            Pos2::new(center.x - rx * t.cos(), center.y + ry * t.sin())
        })
        .collect()
}

/// A non-interactive icon occupying `size`.
pub fn show(ui: &mut egui::Ui, icon: Icon, size: f32, color: Color32) -> Response {
    let (rect, response) = ui.allocate_exact_size(Vec2::splat(size), Sense::hover());
    paint(ui.painter(), rect, icon, color);
    response
}

/// A transparent toolbar button with an icon and optional label. The label
/// (or tooltip) doubles as the accessible name.
pub fn button(ui: &mut egui::Ui, icon: Icon, label: Option<&str>, tooltip: &str) -> Response {
    let enabled = ui.is_enabled();
    let text = label.map(|l| {
        ui.painter().layout_no_wrap(
            l.to_owned(),
            egui::FontId::proportional(12.5),
            theme::SECONDARY,
        )
    });
    let icon_size = 14.0;
    let padding = Vec2::new(7.0, 5.0);
    let width = padding.x * 2.0 + icon_size + text.as_ref().map_or(0.0, |g| g.size().x + 6.0);
    let (rect, response) = ui.allocate_exact_size(Vec2::new(width, 26.0), Sense::click());
    let response = response.on_hover_text(tooltip);
    response.widget_info(|| {
        egui::WidgetInfo::labeled(egui::WidgetType::Button, enabled, label.unwrap_or(tooltip))
    });
    if ui.is_rect_visible(rect) {
        let hovered = response.hovered() && enabled;
        if hovered || response.has_focus() {
            ui.painter().rect_filled(rect, 6, theme::CONTROL_HOVER);
        }
        let color = if !enabled {
            theme::FAINT
        } else if hovered {
            theme::TEXT_STRONG
        } else {
            theme::SECONDARY
        };
        let icon_rect = Rect::from_center_size(
            Pos2::new(rect.left() + padding.x + icon_size / 2.0, rect.center().y),
            Vec2::splat(icon_size),
        );
        paint(ui.painter(), icon_rect, icon, color);
        if let Some(galley) = text {
            let pos = Pos2::new(
                icon_rect.right() + 6.0,
                rect.center().y - galley.size().y / 2.0,
            );
            ui.painter().galley(pos, galley, color);
        }
    }
    response
}
